# Macros

In [1]:
PATH_NAME = '../Data/'
FILE_NAME1 = 'log_NewUsers.log'
FILE_NAME2 = 'log_Purchases.log'

# Importación de librerias

In [2]:
import pandas as pd
import numpy as np
import json
import requests as rq
from datetime import datetime

# Construcción del Dataframe

In [3]:
fn1 = open(PATH_NAME+FILE_NAME1,'r')
js1 = fn1.read()
newusers = pd.read_json(js1)
newusers

client_mobile_device client_mobile_os            datetime ip_country  \
0                   iPad7,5             12.2 2021-01-01 03:13:35         AU   
1         samsung SM-G531BT            5.1.1 2021-01-02 20:30:43         BR   
2               LGE LM-X210            7.1.2 2021-01-04 04:37:39         BR   
3            HUAWEI MYA-U29              6.0 2021-01-05 18:10:40         LK   
4                 iPhone9,3           13.3.1 2021-01-07 13:18:20         AU   
...                     ...              ...                 ...        ...   
102936       HUAWEI STK-LX3               10 2021-02-28 21:07:24         CL   
102937       HUAWEI MRD-LX3                9 2021-02-28 21:48:17         CL   
102938  Mintt ULTRAMINTT X3                9 2021-02-28 22:44:44         AU   
102939     samsung SM-G610M            8.1.0 2021-02-28 23:08:53         CL   
102940       HUAWEI YAL-L21               10 2021-02-28 23:36:16         ZA   

       platform               user_id            version app_code  \
0           ios  gNdE4JGZuXoItNygPSkB  1.17.2-1804790697       st   
1       android  qn0F6bsizLuvnPBF9oj7  1.17.3-1804790685       st   
2       android  UCcF1MeTykbc3Tw3ZMLO  1.17.3-1804790685       st   
3       android  tgI4d8m6C9NZBKSzvKe5  1.17.3-1804790685       st   
4           ios  HFT8rkq8qkRN1bGDRjA5  1.17.2-1804790697       st   
...         ...                   ...                ...      ...   
102936  android  x264D6C0FYLdZPjTyFdl   3.2.3-1804795535       fw   
102937  android  qWBT7gihmGdgUDBfHXEa   3.2.3-1804795535       fw   
102938  android  iGmtqxqjBqEBZCDm0FRS   3.2.3-1804795535       fw   
102939  android  MoDg1M1e3H03Qnt7Qch5   3.2.3-1804795535       fw   
102940  android  NMC3AA7lHPoTZgMkF7yE   3.2.3-1804795535       fw   

             app_name  
0        Slow Traffic  
1        Slow Traffic  
2        Slow Traffic  
3        Slow Traffic  
4        Slow Traffic  
...               ...  
102936  Find The Word  
102937  Find The Word  
102938  Find The Word  
102939  Find The Word  
102940  Find The Word  

[102941 rows x 9 columns]

In [4]:
fn2 = open(PATH_NAME+FILE_NAME2,'r')
js2 = fn2.read()
purchases = pd.read_json(js2)
purchases

datetime  game_basic_level ip_country  order_amount_gross  \
0    2021-01-24 04:05:34                 7         CL           35.446176   
1    2021-01-24 06:00:35                 9         CL            6.569976   
2    2021-01-24 07:04:58                10         CL           29.010654   
3    2021-01-25 04:28:21                12         CL           10.685374   
4    2021-01-25 04:43:59                12         CL            5.367659   
...                  ...               ...        ...                 ...   
2856 2021-01-17 22:48:13                10         VE           22.575198   
2857 2021-01-17 23:13:31                10         VE            6.028233   
2858 2021-02-05 13:03:47                 8         ID           33.966516   
2859 2021-02-25 13:55:32                 7         AU           38.008819   
2860 2021-02-27 00:23:41                 8         AU           33.459975   

     order_payment_provider  order_transaction_id platform  \
0                    Itunes            1184529158      ios   
1                    Itunes             774979105      ios   
2                    Itunes            7793447456      ios   
3                    Itunes            3111804466      ios   
4                    Itunes            4652947566      ios   
...                     ...                   ...      ...   
2856                 Itunes            9465335020      ios   
2857                 Itunes            5453471344      ios   
2858             GooglePlay            7487755476  android   
2859             GooglePlay            3429525576  android   
2860             GooglePlay            9099537065  android   

                   user_id            version app_code      app_name currency  
0     23zBzzPT2thJHJb4ZMwl  1.17.4-1804792971       st  Slow Traffic      EUR  
1     23zBzzPT2thJHJb4ZMwl  1.17.4-1804792971       st  Slow Traffic      CHF  
2     23zBzzPT2thJHJb4ZMwl  1.17.4-1804792971       st  Slow Traffic      USD  
3     23zBzzPT2thJHJb4ZMwl  1.17.4-1804792971       st  Slow Traffic      USD  
4     23zBzzPT2thJHJb4ZMwl  1.17.4-1804792971       st  Slow Traffic      GBP  
...                    ...                ...      ...           ...      ...  
2856  0LocGHQTQP7NngKsIqSz   2.6.1-1804789113       dd   Dark Diving  unknown  
2857  0LocGHQTQP7NngKsIqSz   2.6.1-1804789113       dd   Dark Diving      CHF  
2858  rcRU3r0I22OyEFYAK5XS   2.7.0-1804792974       dd   Dark Diving      CHF  
2859  jhAUkMKU0vJTLGzfZ6Ij   2.7.0-1804792974       dd   Dark Diving      CHF  
2860  jhAUkMKU0vJTLGzfZ6Ij   2.7.0-1804792974       dd   Dark Diving      CHF  

[2861 rows x 12 columns]

# Analisis de datos

<b> PARA LOS NUEVOS USUARIOS

In [5]:
newusers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102941 entries, 0 to 102940
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   client_mobile_device  102941 non-null  object        
 1   client_mobile_os      102941 non-null  object        
 2   datetime              102941 non-null  datetime64[ns]
 3   ip_country            102941 non-null  object        
 4   platform              102941 non-null  object        
 5   user_id               102941 non-null  object        
 6   version               102941 non-null  object        
 7   app_code              102941 non-null  object        
 8   app_name              102941 non-null  object        
dtypes: datetime64[ns](1), object(8)
memory usage: 7.1+ MB


In [6]:
newusers.nunique()

client_mobile_device      3493
client_mobile_os           105
datetime                101634
ip_country                  67
platform                     2
user_id                 102941
version                     48
app_code                     3
app_name                     3
dtype: int64

In [7]:
newusers.describe()

<ipython-input-7-6d5ad4e79450>:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  newusers.describe()


client_mobile_device client_mobile_os             datetime ip_country  \
count                102941           102941               102941     102941   
unique                 3493              105               101634         67   
top        samsung SM-A105M                9  2021-02-28 06:08:37         CL   
freq                   1843            32275                    3      46808   
first                   NaN              NaN  2021-01-01 00:00:53        NaN   
last                    NaN              NaN  2021-02-28 23:59:55        NaN   

       platform               user_id           version app_code  \
count    102941                102941            102941   102941   
unique        2                102941                48        3   
top     android  RKX45WWymSLRpBwGam23  3.2.3-1804795535       fw   
freq      80720                     1             14876    68906   
first       NaN                   NaN               NaN      NaN   
last        NaN                   NaN               NaN      NaN   

             app_name  
count          102941  
unique              3  
top     Find The Word  
freq            68906  
first             NaN  
last              NaN

In [8]:
newusers.isna().sum()

client_mobile_device    0
client_mobile_os        0
datetime                0
ip_country              0
platform                0
user_id                 0
version                 0
app_code                0
app_name                0
dtype: int64

In [9]:
newusers.duplicated().any()

False

In [10]:
newusers['platform'].value_counts()

android    80720
ios        22221
Name: platform, dtype: int64

<b> PARA LOS COMPRAS REALIZADAS

In [11]:
purchases.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2861 entries, 0 to 2860
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   datetime                2861 non-null   datetime64[ns]
 1   game_basic_level        2861 non-null   int64         
 2   ip_country              2861 non-null   object        
 3   order_amount_gross      2861 non-null   float64       
 4   order_payment_provider  2861 non-null   object        
 5   order_transaction_id    2861 non-null   int64         
 6   platform                2861 non-null   object        
 7   user_id                 2861 non-null   object        
 8   version                 2861 non-null   object        
 9   app_code                2861 non-null   object        
 10  app_name                2861 non-null   object        
 11  currency                2861 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(2), obje

In [12]:
purchases.nunique()

datetime                  2857
game_basic_level           602
ip_country                  27
order_amount_gross        2861
order_payment_provider       2
order_transaction_id      2861
platform                     3
user_id                    699
version                     23
app_code                     3
app_name                     3
currency                     5
dtype: int64

In [13]:
purchases.describe()

game_basic_level  order_amount_gross  order_transaction_id
count       2861.000000         2861.000000          2.861000e+03
mean         176.882209           26.310582          5.026741e+09
std          197.306477           34.032660          2.837513e+09
min            2.000000            0.063427          2.400889e+06
25%           26.000000            8.222154          2.589577e+09
50%          103.000000           19.512636          5.047037e+09
75%          263.000000           33.889809          7.439688e+09
max         1427.000000          775.862223          9.999436e+09

In [14]:
purchases.isna().sum()

datetime                  0
game_basic_level          0
ip_country                0
order_amount_gross        0
order_payment_provider    0
order_transaction_id      0
platform                  0
user_id                   0
version                   0
app_code                  0
app_name                  0
currency                  0
dtype: int64

In [15]:
purchases.duplicated().any()

False

In [16]:
purchases['platform'].value_counts() # Se debe hacer limpieza de valores desconocidos

ios        1607
android    1191
             63
Name: platform, dtype: int64

In [17]:
purchases['currency'].value_counts() # Se debe hacer limpieza de valores desconocidos

EUR        807
CHF        772
GBP        585
unknown    426
USD        271
Name: currency, dtype: int64

# Limpieza y tratamiento de datos

In [18]:
purchases['platform'] = purchases['platform'].replace({'':'ios'})
purchases['platform'].value_counts() # Eliminamos los 'unknown' del campo 'platform'

ios        1670
android    1191
Name: platform, dtype: int64

In [19]:
purchases['currency'] = purchases['currency'].replace({'unknown':'USD'})
purchases['currency'].value_counts() # Eliminamos los 'unknown' del campo 'currency'

EUR    807
CHF    772
USD    697
GBP    585
Name: currency, dtype: int64

# Rescate de divisas desde la api

In [20]:
date = '2021-03-19'
currency = 'USD'
a = rq.get('https://api.exchangeratesapi.io/' + date + '?symbols=' + currency)
b= a.json()
c = pd.DataFrame(b)

print(c['rates'][currency])

1.1891


In [21]:
ratios = []
for i in range(len(purchases)):
    if purchases['currency'][i]=='EUR':
        b = 1
    else:
        ratio = rq.get('https://api.exchangeratesapi.io/' + str(purchases['datetime'][i]).split(' ')[0] + '?symbols=' + purchases['currency'][i])
        a = ratio.json()
        b = a['rates'][purchases['currency'][i]]
    ratios.append(b)
print(ratios)

[1, 1.0773, 1.2158, 1.2152, 0.888, 0.88698, 0.88395, 1.0759, 0.88603, 1, 1, 1.0783, 1.0798, 1.2136, 1.0798, 0.88383, 1.0798, 1, 1.0798, 1, 1.2136, 1.0798, 1.2136, 1.0798, 1, 1.2084, 0.882, 1.0816, 1.0816, 0.88145, 1.2017, 1.1996, 1.0818, 1.0825, 1.0825, 1.1983, 1, 1, 1, 0.87833, 0.87833, 1.2025, 1, 1, 1, 1.0817, 1.0817, 1, 1.0811, 1.2296, 1, 0.9019, 1, 1.0827, 1.0827, 1, 1.225, 1, 1, 0.90235, 1, 1.0838, 1.2166, 1, 1.0805, 1.2124, 1, 0.89045, 1.2158, 1.2158, 1, 1.2091, 1, 0.88383, 1.2136, 1.0798, 1.2136, 1, 1, 0.87538, 1, 1.1983, 0.87538, 1, 0.8765, 0.87755, 1.2108, 1, 0.86508, 1, 0.90235, 0.90235, 1, 1.2163, 0.90235, 0.90235, 0.90235, 1, 1.0838, 1.0838, 1, 1, 1, 1, 1.0812, 0.8944, 1.0812, 0.8944, 1.2161, 1, 1, 1.2166, 1.2123, 1.2123, 1.077, 1, 1.2123, 1.077, 1.077, 1.2064, 1.2064, 1.0747, 1.2101, 0.88563, 1.2158, 1.2158, 1.0773, 1.2152, 0.88698, 1.0789, 0.88395, 0.88383, 1.0798, 1.0769, 1, 1.2161, 1, 1.2123, 1.077, 1, 1, 1.08, 1.2143, 1, 1, 1, 1, 1.2025, 1.0842, 0.87833, 1, 1, 1.2108, 

In [22]:
purchases['ratios'] = ratios
print(purchases)

                datetime  game_basic_level ip_country  order_amount_gross  \
0    2021-01-24 04:05:34                 7         CL           35.446176   
1    2021-01-24 06:00:35                 9         CL            6.569976   
2    2021-01-24 07:04:58                10         CL           29.010654   
3    2021-01-25 04:28:21                12         CL           10.685374   
4    2021-01-25 04:43:59                12         CL            5.367659   
...                  ...               ...        ...                 ...   
2856 2021-01-17 22:48:13                10         VE           22.575198   
2857 2021-01-17 23:13:31                10         VE            6.028233   
2858 2021-02-05 13:03:47                 8         ID           33.966516   
2859 2021-02-25 13:55:32                 7         AU           38.008819   
2860 2021-02-27 00:23:41                 8         AU           33.459975   

     order_payment_provider  order_transaction_id platform  \
0            

In [23]:
purchases['order_amount_gross_in_EUR'] = purchases['order_amount_gross']/purchases['ratios']
print(purchases)

                datetime  game_basic_level ip_country  order_amount_gross  \
0    2021-01-24 04:05:34                 7         CL           35.446176   
1    2021-01-24 06:00:35                 9         CL            6.569976   
2    2021-01-24 07:04:58                10         CL           29.010654   
3    2021-01-25 04:28:21                12         CL           10.685374   
4    2021-01-25 04:43:59                12         CL            5.367659   
...                  ...               ...        ...                 ...   
2856 2021-01-17 22:48:13                10         VE           22.575198   
2857 2021-01-17 23:13:31                10         VE            6.028233   
2858 2021-02-05 13:03:47                 8         ID           33.966516   
2859 2021-02-25 13:55:32                 7         AU           38.008819   
2860 2021-02-27 00:23:41                 8         AU           33.459975   

     order_payment_provider  order_transaction_id platform  \
0            

# Obtener metricas solicitadas

<b> Volumen total de ingresos generados

In [24]:
VolumenTotalIngresos = purchases['order_amount_gross_in_EUR'].sum()
print('Volumen total de ingresos generados es de €',round(VolumenTotalIngresos,3))

Volumen total de ingresos generados es de € 72530.946


<b> Número total de pagos

In [25]:
NumeroTotalPagos = purchases['order_amount_gross_in_EUR'].count()
print('Número total de pagos es de',NumeroTotalPagos)

Número total de pagos es de 2861


<b> Número total de jugadores

In [26]:
NumeroTotalPlayers = newusers['user_id'].nunique()
print('Número total de jugadores',NumeroTotalPlayers)

Número total de jugadores 102941


<b> Número total de jugadores que han realizado alguna compra

In [27]:
NumeroTotalPlayersBuyers = purchases['user_id'].nunique()
print('Número total de jugadores que han realizado alguna compra es de',NumeroTotalPlayersBuyers)

Número total de jugadores que han realizado alguna compra es de 699


<b> Promedio de ingresos generados por jugador

In [28]:
PromedioIngresosPorPlayer = VolumenTotalIngresos/NumeroTotalPlayers
print('€',PromedioIngresosPorPlayer)

€ 0.7045875402521005


<b> Promedio de pagos por jugador

In [29]:
TotalPagos = purchases['order_transaction_id'].nunique()
PromedioPagosXPlayers = TotalPagos/NumeroTotalPlayers
print('Promedio de pagos por jugador es de',round(PromedioPagosXPlayers,4))

Promedio de pagos por jugador es de 0.0278


<b> Promedio de ingresos por comprador (jugadores que hayan realizado al menos
una compra) agrupado por el país desde el que se registraron.

In [37]:
PromedioIngresosXBuyersXPais = purchases['order_amount_gross_in_EUR'].groupby(purchases['ip_country']).mean()
print('Promedio de ingresos por comprador agrupado por el país desde el que se registraron es')
print(PromedioIngresosXBuyersXPais)

Promedio de ingresos por comprador agrupado por el país desde el que se registraron es
ip_country
AR    29.753230
AU    26.226561
BN    17.995454
BR    28.201268
CL    23.556350
CO    57.357492
CR    17.302790
GA     3.855741
GF    36.157822
GH    25.109070
ID    35.388869
LK    11.038784
LR    19.664330
MQ    11.882223
MV     8.681378
MY    30.275891
MZ    14.566634
NG    13.486017
NZ    20.042938
PA    69.249227
PF     7.730589
PH     9.290955
PY     8.448877
VE    23.497854
VN     7.521299
ZA    18.081017
ZW    40.448744
Name: order_amount_gross_in_EUR, dtype: float64


<b> Ingresos totales por día de registro. Únicamente para aquellos usuarios
registrados en febrero de 2021.

In [31]:
def ingtot():
    uid = [] # Usuarios registrados en febrero del 2021
    date = [] # Fecha en la que se inscribieron esos usuarios
    for i in range(len(newusers)): # Busca en el DataFrame "newusers" la fecha de registro y el ID de los usuarios
        if newusers['datetime'][i]> datetime.strptime('2021-02-01', '%Y-%m-%d') and newusers['datetime'][i]< datetime.strptime('2021-03-01', '%Y-%m-%d'):
            uid.append(newusers['user_id'][i])
            date.append(str(newusers['datetime'][i]).split(' ')[0])
        else:
            pass
    df = pd.DataFrame(list(zip(date,uid)), columns = ['Fecha','user_id']) # Se genera un DataFrame con las listas anteriores
    date1 = []
    ing = []
    userbuy = []
    for e in range(len(purchases)): # Se busca las compras realizadas por los usuario de febrero en el dia que ingresaron como usuarios
        if purchases['user_id'][e] in list(df['user_id']):
            if str(purchases['datetime'][e]).split(' ')[0]==str(df[df['user_id']==purchases['user_id'][e]]['Fecha'].any()):
                date1.append(str(df[df['user_id']==purchases['user_id'][e]]['Fecha'].any()))
                ing.append(purchases['order_amount_gross_in_EUR'][e])
                userbuy.append(purchases['user_id'][e])
            else:
                pass
        else:
            pass
        
    df2 = pd.DataFrame(date1, columns = ['Fecha'])
    df2['user_id'] = userbuy
    df2['ingresos'] = ing
    return df2

UsuariosconcomprasenFebrero = ingtot()
print(UsuariosconcomprasenFebrero)

          Fecha               user_id   ingresos
0    2021-02-07  boxkcx7vER3ZQhpYYuah  53.543271
1    2021-02-07  boxkcx7vER3ZQhpYYuah  15.673896
2    2021-02-17  FVYnZC1SHdffBzfgowtF  20.156926
3    2021-02-28  XdT3YB9GYIuuULtsangz  29.033779
4    2021-02-16  bZmXdSwILwOYuk3m4egi  58.200744
..          ...                   ...        ...
141  2021-02-24  SwkEz8u5j5nK9bwGSb1R  57.152298
142  2021-02-24  SwkEz8u5j5nK9bwGSb1R  28.385011
143  2021-02-26  KthoLCUTfjS41YHC8PPP   2.569758
144  2021-02-28  Onw81nTHwnaR9sWfmSrx  25.119111
145  2021-02-05  rcRU3r0I22OyEFYAK5XS  31.377844

[146 rows x 3 columns]


In [39]:
CompraXDiaRegistro = UsuariosconcomprasenFebrero.groupby(UsuariosconcomprasenFebrero['Fecha']).sum()
print('Ingresos totales por día de registro para los usuarios registrados en febrero del 2021 fue de')
print(CompraXDiaRegistro)

Ingresos totales por día de registro para los usuarios registrados en febrero del 2021 fue de
              ingresos
Fecha                 
2021-02-01  129.129592
2021-02-02   27.182866
2021-02-03   34.865919
2021-02-04   69.841508
2021-02-05   79.522613
2021-02-06  278.408342
2021-02-07  148.962124
2021-02-08   28.564378
2021-02-09   95.026768
2021-02-10  101.958470
2021-02-11   82.108670
2021-02-12   53.847163
2021-02-13   15.481143
2021-02-14   70.962880
2021-02-16  233.777641
2021-02-17  114.694719
2021-02-18   22.752718
2021-02-19  721.299186
2021-02-20   43.919684
2021-02-21  151.309668
2021-02-22  191.647556
2021-02-23   80.616251
2021-02-24  114.310566
2021-02-25    6.460417
2021-02-26   69.335928
2021-02-27  166.000624
2021-02-28  421.954990


# Conclusiones

1. Se observa que el dia con mayor consumo por registro de usuario en febrero del 2021 fue el dia 19 de Febrero con €721.3.
2. El pais con mayor ingreso por compradores es Panama.
3. De un total de 102941 usuarios, solamente 699 son compradores siendo estos el 0.68% del total de usuarios.